controllers/cloud_config_sync_controller.go

```diff
@@ -0,0 +1,158 @@
+			return ctrl.Result{}, err
+		}
+		sourceCM, err = r.prepareSourceConfigMap(sourceCM, infra)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	} else if err != nil {
+		klog.Errorf("unable to get managed cloud-config for sync")
+		return ctrl.Result{}, err
+	}
+
+	targetCM := &corev1.ConfigMap{}
+	targetConfigMapKey := client.ObjectKey{
+		Namespace: r.TargetNamespace,
+		Name:      cloudConfigMapName,
+	}
+
+	if err := r.Get(ctx, targetConfigMapKey, targetCM); err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("unable to get target cloud-config for sync")
+		return ctrl.Result{}, err
+	}
+
+	if r.isCloudConfigEqual(sourceCM, targetCM) {
+		klog.Infof("source and target cloud-config content are equal, no sync needed")
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.syncCloudConfigData(ctx, sourceCM, targetCM); err != nil {
+		klog.Errorf("unable to sync cloud config")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *CloudConfigReconciler) prepareSourceConfigMap(source *corev1.ConfigMap, infra *configv1.Infrastructure) (*corev1.ConfigMap, error) {
+	// Keys might be different between openshift-config/cloud-config and openshift-config-managed/kube-cloud-config
+	// Always use "cloud.conf" which is default one across openshift
+	cloudConfCm := source.DeepCopy()
+	if _, ok := cloudConfCm.Data[defaultConfigKey]; ok {
+		return cloudConfCm, nil
+	}
+
+	infraConfigKey := infra.Spec.CloudConfig.Key
+	if val, ok := cloudConfCm.Data[infraConfigKey]; ok {
+		cloudConfCm.Data[defaultConfigKey] = val
+		delete(cloudConfCm.Data, infraConfigKey)
+		return cloudConfCm, nil
+	}
+	return nil, fmt.Errorf(
+		"key %s specified in infra resource does not found in source configmap %s",
+		infraConfigKey, client.ObjectKeyFromObject(source),
+	)
+}
+
+func (r *CloudConfigReconciler) isCloudConfigEqual(source *corev1.ConfigMap, target *corev1.ConfigMap) bool {
+	return source.Immutable == target.Immutable &&
+		reflect.DeepEqual(source.Data, target.Data) && reflect.DeepEqual(source.BinaryData, target.BinaryData)
+}
+
+func (r *CloudConfigReconciler) syncCloudConfigData(ctx context.Context, source *corev1.ConfigMap, target *corev1.ConfigMap) error {
+
+	target.SetName(cloudConfigMapName)
+	target.SetNamespace(r.TargetNamespace)
+	target.Data = source.Data
+	target.BinaryData = source.BinaryData
+	target.Immutable = source.Immutable
+
+	// check if target config exists, create if not
+	err := r.Get(ctx, client.ObjectKeyFromObject(target), &corev1.ConfigMap{})
+
+	if err != nil && errors.IsNotFound(err) {
+		return r.Create(ctx, target)
+	} else if err != nil {
+		return err
+	}
+
+	return r.Update(ctx, target)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *CloudConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	build := ctrl.NewControllerManagedBy(mgr).
+		For(
+			&corev1.ConfigMap{},
+			builder.WithPredicates(ownCloudConfigPredicate(r.TargetNamespace)),
```

Comment:
I think there's a way you can `Or` predicates, which would mean you don't need the first set of `Watches` below